### PR TITLE
[backend] engine.ts

### PR DESCRIPTION
**Pull request by JeremyCloarec**
### Proposed changes

* engine.ts

### Related issues

*
*

### Checklist

- [x] I consider the submitted work as finished
- [ ] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments

It's far from perfect, but a first step towards making our engine layer more robust and maintainable.
A lot of things are not properly typed or typed as `any`, and still need improvements.
Next steps would be:
* split engine concerns into seperate files: raw engine layer, engine query builder & filter handling, platform init & mapping etc...
* improve types in current engine.ts: create a type for es body & filters etc...
* move remaining js files to ts: data-builder, data-initialization etc...

### CODE REVIEW

**Review by marieflorescontact — COMMENTED** [review]
looks good functionnaly, I can't see anything broken. 
I haven't look at the code yet 😅

**Review by MTorbay-Filigran — COMMENTED** [review]
Great job! I mainly reviewed the typing, and everything looks good to me.
I noticed that return values for some functions are explicitly typed while others are not, is there a specific reason for that?

For functional testing, do you have a few use cases in mind that we could try to make sure nothing is broken?

**Comment by JeremyCloarec** [comment]
> Great job! I mainly reviewed the typing, and everything looks good to me. I noticed that return values for some functions are explicitly typed while others are not, is there a specific reason for that?
> 
> For functional testing, do you have a few use cases in mind that we could try to make sure nothing is broken?

The fact that some functions are not explicitly typed was just to save some time to be honest 😅 Ideally every return values should be explicitly typed to improve code readability. But that's something that we'll be able to improve in the future now.

As for functional testing, I'd say the things that you can test is making sure that complex filters are still all working properly. Engine is used in almost all queries, so it's already pretty well covered with our test suite, but it could be a great idea to test edge cases manually: making sure that pagination works properly when moving to the next page/going back to the next page, deleting a large item, complex filters etc...

File: opencti-platform/opencti-graphql/src/database/draft-engine.ts [hunk 1]
```diff
@@ -62,15 +64,17 @@ const elRemoveCreateElementFromDraft = async (context, user, element) => {
...
+const elRemoveUpdateElementFromDraft = async (context: AuthContext, user: AuthUser, element: BasicStoreCommon): Promise<void> => {
   if (element.draft_change?.draft_operation !== DRAFT_OPERATION_UPDATE) {
     return;
   }
 
   // apply reverse field patch
   const elementWithRefs = await storeLoadByIdWithRefs(context, user, element.internal_id);
-  const reverseUpdateFieldPatch = buildReverseUpdateFieldPatch(element.draft_change.draft_updates_patch);
-  await updateAttributeFromLoadedWithRefs(context, user, elementWithRefs, reverseUpdateFieldPatch);
+  if (element.draft_change.draft_updates_patch) {
```
Comment (SarahBocognano) [inline opencti-platform/opencti-graphql/src/database/draft-engine.ts:74]: Why put a condition here ? Was it missing ?

File: opencti-platform/opencti-graphql/src/database/draft-engine.ts [hunk 2]
```diff
@@ -237,12 +248,12 @@ export const elDeleteDraftContextFromUsers = async (context, user, draftId) => {
         }
       },
     },
-  }).catch((err) => {
+  }).catch((err: any) => {
```
Comment (SarahBocognano) [inline opencti-platform/opencti-graphql/src/database/draft-engine.ts:251]: Not possible to have a typed error here ? (curious)

File: opencti-platform/opencti-graphql/src/modules/draftWorkspace/draftWorkspace-domain.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ export const getObjectsCount = async (context: AuthContext, user: AuthUser, draf
     convertEntityTypeLabel: true,
   };
   const draftContext = { ...context, draft_context: draft.id };
-  const distributionResult = await elAggregationCount(draftContext, context.user, READ_INDEX_DRAFT_OBJECTS, opts);
```
Comment (SarahBocognano) [inline opencti-platform/opencti-graphql/src/modules/draftWorkspace/draftWorkspace-domain.ts:61]: Out of curiosity: why remove the context ?

File: opencti-platform/opencti-graphql/src/types/store.d.ts [hunk 4]
```diff
@@ -66,7 +66,8 @@ interface Representative {
 interface InternalEditInput {
   key: string;
   operation?: EditOperation | null;
-  value: (string | object | null)[];
+  value: any[];
```
Comment (SarahBocognano) [inline opencti-platform/opencti-graphql/src/types/store.d.ts]: why changing the value type by any[] ?

File: opencti-platform/opencti-graphql/src/database/draft-engine.ts (hunk 1 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/database/draft-engine.ts:74]: It wasn't really missing, but since I typed the buildReverseUpdateFieldPatch input as string, I needed to make sure that element.draft_change.draft_updates_patch wasn't undefined/null here. This also prevents calling updateAttributeFromLoadedWithRefs when it's not needed

File: opencti-platform/opencti-graphql/src/database/draft-engine.ts (hunk 2 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/database/draft-engine.ts:251]: It's probably possible, but a bit hard to do. We'd need to know what kind of error deleteByQuery can raise, both for opensearch and elastic clients, and I didn't take the time to dive into it. But it'd be more clean to do so you're right :)

File: opencti-platform/opencti-graphql/src/modules/draftWorkspace/draftWorkspace-domain.ts (hunk 3 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/modules/draftWorkspace/draftWorkspace-domain.ts:61]: Mostly to solve a TS issue: the user in AuthContext is typed as possibly undefined, so using context.user was throwing a TS issue since elAggregationCount doesn't accept an undefined user.
I don't know why we were using context.user instead of user directly here though

File: opencti-platform/opencti-graphql/src/types/store.d.ts (hunk 4 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/types/store.d.ts]: I was getting a lot of TS errors with the (string | object | null)[] typing, so this was the shortcut I took to make it work.
But that's a good remark, I pushed a new commit to revert to a better typing than any[]! And I applied the needed changes to handle properly the typing

**Review by MTorbay-Filigran — APPROVED** [review]
I tested functionnaly with the use cases you gave me and found nothing wrong so I validate :)

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 5]
```diff
@@ -2517,62 +2552,29 @@ export const buildLocalMustFilter = (validFilter) => {
...
   if (arrayKeys.length > 1) {
     throw UnsupportedError('A filter with these multiple keys is not supported', { keys: arrayKeys });
   }
   // at this point arrayKeys === ['entity_type']
 
   // we'll build these new filters or filterGroup, depending on the situation
-  let newFilter;
-  let newFilterGroup;
+  let newFilter: Filter | undefined;
+  let newFilterGroup: FilterGroup = emptyFilterGroup;
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]:
```suggestion
  let newFilterGroup: FilterGroup | undefined;
```
avoid adding a useless empty filter group in the array filters.filterGroups....

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 6]
```diff
@@ -2592,78 +2594,708 @@ const adaptFilterToEntityTypeFilterKey = (filter) => {
...
+  if (arrayKeys[0] !== IDS_FILTER || arrayKeys.length > 1) {
+    throw UnsupportedError('A filter with these multiple keys is not supported', { keys: arrayKeys });
+  }
+  if (mode === 'and') {
+    throw UnsupportedError('Unsupported filter: \'And\' operator between values of a filter with key = \'ids\' is not supported');
+  }
+  // at this point arrayKey === ['ids'], and mode is always 'or'
+
+  // we'll build these new filters or filterGroup, depending on the situation
+  let newFilterGroup: FilterGroup = emptyFilterGroup;
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]: same

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 7]
```diff
@@ -2592,78 +2594,708 @@ const adaptFilterToEntityTypeFilterKey = (filter) => {
...
-  // at this point arrayKey === ['ids'], and mode is always 'or'
-
-  // we'll build these new filters or filterGroup, depending on the situation
-  let newFilterGroup;
-
-  const idsArray = [...ids_attributes]; // the keys to handle additionally
-
-  if (operator === 'nil' || operator === 'not_nil') { // nil and not_nil operators must have a single key
-    const filters = idsArray.map((idKey) => { return { ...filter, key: idKey }; });
+  let newFilterGroup: FilterGroup = emptyFilterGroup;
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]: same for every adaptXXFilter functions

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 8]
```diff
@@ -2809,36 +3441,37 @@ const adaptFilterToPirFilterKeys = async (context, user, filterKey, filter) => {
...
     key: ['metrics'],
-    mode: 'and',
+    mode: FilterMode.And,
     nested: [
-      { key: 'name', values: [filter.key], operator: 'eq' },
+      { key: 'name', values: [filter.key], operator: FilterOperator.Eq },
       { key: 'value', values: filter.values, operator: filter.operator, mode: filter.mode },
-    ]
+    ],
+    values: []
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts:3453]: values is useless for a filter with a 'nested' attribute. I guess it's a question of typing...

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 9]
```diff
@@ -3066,14 +3640,14 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
-          keys.forEach((relKey) => {
+            : types.map((t: any) => buildRefRelationKey(t, '*'));
+          keys.forEach((relKey: any) => {
             regardingFilters.push({ key: [relKey], operator: filter.operator, values: ['EXISTS'] });
           });
         } else {
           const keys = isEmptyField(types)
             ? buildRefRelationKey('*', '*')
-            : types.flatMap((t) => [buildRefRelationKey(t, ID_INTERNAL), buildRefRelationKey(t, ID_INFERRED)]);
+            : types.flatMap((t: any) => [buildRefRelationKey(t, ID_INTERNAL), buildRefRelationKey(t, ID_INFERRED)]);
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]:
```suggestion
            : types.flatMap((t: string) => [buildRefRelationKey(t, ID_INTERNAL), buildRefRelationKey(t, ID_INFERRED)]);
```

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 10]
```diff
@@ -3066,14 +3640,14 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
         const mode = (filter.operator === 'eq' || isEmptyField(filter.operator)) ? FilterMode.Or : FilterMode.And;
         if (isEmptyField(ids)) {
           const keys = isEmptyField(types) ? buildRefRelationKey('*', '*')
-            : types.map((t) => buildRefRelationKey(t, '*'));
-          keys.forEach((relKey) => {
+            : types.map((t: any) => buildRefRelationKey(t, '*'));
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]:
```suggestion
            : types.map((t: string) => buildRefRelationKey(t, '*'));
```

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 11]
```diff
@@ -3066,14 +3640,14 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
...
+          keys.forEach((relKey: any) => {
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]:
```suggestion
          keys.forEach((relKey: string) => {
```

File: opencti-platform/opencti-graphql/src/database/engine.ts [hunk 12]
```diff
@@ -3262,174 +3836,66 @@ const completeSpecialFilterKeys = async (context, user, inputFilters) => {
       if (arrayKeys.length > 1) {
         throw UnsupportedError('A filter with these multiple keys is not supported', { keys: arrayKeys });
       }
-      const definition = schemaAttributesDefinition.getAttributeByName(key[0]);
+      const definition: any = schemaAttributesDefinition.getAttributeByName(key[0]);
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/database/engine.ts]:
```suggestion
      const definition: AttributeDefinition = schemaAttributesDefinition.getAttributeByName(key[0]);
```

File: opencti-platform/opencti-graphql/src/database/engine.ts (hunk 8 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/database/engine.ts:3453]: Yes, the idea was to stick to the Filter type that always requires a values attribute

File: opencti-platform/opencti-graphql/src/database/engine.ts (hunk 12 above)
Comment (JeremyCloarec) [reply opencti-platform/opencti-graphql/src/database/engine.ts]: I changed it to type it as ComplexAttribute instead, since AttributeDefinition doesn't always contain a format

Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/database/engine.ts]: nice ! perfect

File: opencti-platform/opencti-graphql/src/database/engine.ts (hunk 8 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/database/engine.ts:3453]: okay I understand
